pandas/core/base.py

```diff
@@ -584,7 +585,7 @@ def _shallow_copy(self, obj, **kwargs):
         return self._constructor(obj, **kwargs)
 
 
-class IndexOpsMixin:
+class IndexOpsMixin(Generic[ArrayLike]):
```

Comment:
Hmm I wouldn't really consider a Mixin to be a Generic class - have you seen this pattern used elsewhere?

Reply:
> Hmm I wouldn't really consider a Mixin to be a Generic class

see errors in OP of #31518

>pandas/core/indexes/base.py:246: error: Type variable "pandas._typing.ArrayLike" is unbound
pandas/core/indexes/base.py:246: note: (Hint: Use "Generic[ArrayLike]" or "Protocol[ArrayLike]" base class to bind "ArrayLike" inside a class)
pandas/core/indexes/base.py:246: note: (Hint: Use "ArrayLike" in function signature to bind "ArrayLike" inside a function)

The goal here is to use the ArrayLike alias from pandas._typing to replace uses of `Union[ExtensionArray, np.ndarray]`. ArrayLike is a typevar and hence mypy reports errors if the typevar is not bound.

In Index we want the return type of _values to be the same type as self._data. The way to bind typevars within a class is to use Generic.

More value would come from having the correct type available in calling code instead of a Union type to reduce casts, asserts and type ignores to perform narrowing.

I'll mark this as draft till this is needed.

Reply:
Thanks. So my question was more around the general approach of dealing with Mixins here. The mypy suggested approach I think is using Protocol:

https://mypy.readthedocs.io/en/stable/more_types.html#mixin-classes

So just curious if you found this approach elsewhere in the docs or are using another project as a precedent

Reply:
>The mypy suggested approach I think is using Protocol:
>
>https://mypy.readthedocs.io/en/stable/more_types.html#mixin-classes

from those docs

> one can define a protocol that defines common functionality for host classes instead of adding required abstract methods to every mixin

At the moment `_values` has already been added (not in this PR) as an abstract method on L680 so changing to Protocol is orthogonal to this PR.

PEP 544 states

>Programmers are free to not use them even if they use type annotations.
>There is no intent to make protocols non-optional in the future.

we can have the discussion regarding the use of protocol and the migration path in #33274

This PR is about binding Typvars.

In common with the Generic abc, Protocol is generic and would still need the typevar in the (Protocol) definition class...
(example from PEP 544)

>class Box(Protocol[T_co]):
    def content(self) -> T_co:
        ...

so if we adopt Protocols in the future we would use `SomeClassWithJustTheAbstractMethods[Protocol[ArrayLike]]` instead of `IndexOpsMixin(Generic[ArrayLike])` with the abstract methods included.


> So my question was more around the general approach of dealing with Mixins here

hopefully answered as not relevant.

>So just curious if you found this approach elsewhere in the docs or are using another project as a precedent

following advice from mypy `pandas/core/indexes/base.py:246: note: (Hint: Use "Generic[ArrayLike]" or "Protocol[ArrayLike]" base class to bind "ArrayLike" inside a class)`